### PR TITLE
Backwards euler models

### PR DESCRIPTION
**Pull request by MauriceHendrix**
## Description

This PR implements backwards Euler models

## Motivation and Context

BE is one of the required model types to be able to replace pycml

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have updated all documentation necessary.
- [X] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically 
Also tested (and working) in chaste test

### CODE REVIEW

File: chaste_codegen/tests/test_console_script.py [hunk 1]
```diff
@@ -112,7 +112,7 @@ def test_script_class_convtype_output_dll_loadable(capsys, tmp_path):
                                    os.path.join(tmp_path, 'output_class.c'))
 
 
-def test_script_output_expose_annotated_variables(capsys, tmp_path):
+def test_script_outputexp_ose_annotated_variables(capsys, tmp_path):
```
Comment (jonc125) [inline chaste_codegen/tests/test_console_script.py]: Find & replace!

File: chaste_codegen/tests/test_codegen.py [hunk 2]
```diff
@@ -120,7 +137,27 @@ def test_dymaic_cvode(tmp_path):
                                                expected_cpp_path)
 
 
-def test_expose_annotated_variables(tmp_path):
+def test_dymaic_BE(tmp_path):
```
Comment (jonc125) [inline chaste_codegen/tests/test_codegen.py]: test_dymaic_BE -> test_dynamic_BE

File: chaste_codegen/tests/test_codegen.py [hunk 3]
```diff
@@ -120,7 +137,27 @@ def test_dymaic_cvode(tmp_path):
...
+    expected_hpp_path = \
+        os.path.join(cg.DATA_DIR, 'tests', 'chaste_reference_models', 'BE', 'dynamic_luo_rudy_1994.hpp')
+    expected_cpp_path = \
+        os.path.join(cg.DATA_DIR, 'tests', 'chaste_reference_models', 'BE', 'dynamic_luo_rudy_1994.cpp')
+    # Compare against reference
+    test_utils.compare_model_against_reference('BE', chaste_model, tmp_path, expected_hpp_path,
+                                               expected_cpp_path)
+
+
+def testexp_ose_annotated_variables(tmp_path):
```
Comment (jonc125) [inline chaste_codegen/tests/test_codegen.py]: Find & replace!

File: chaste_codegen/data/tests/cronjob_reference_models/Opt/desktop.ini [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[LocalizedFileNames]
```
Comment (jonc125) [inline chaste_codegen/data/tests/cronjob_reference_models/Opt/desktop.ini]: Should this file be in the repo?

File: chaste_codegen/templates/be_model.cpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+{% include "Shared/cpp/UseCellMLDefaultStimulus" %}
+{% include "Shared/cpp/GetIntracellularCalciumConcentration" %}
+{% include "BE/cpp/constructor" %}
+{% include "Shared/cpp/destructor" %}
+{% include "Shared/cpp/VerifyStateVariables" %}
+{% include "Shared/cpp/GetIIonic" %}
+{% include "BE/cpp/ComputeResidual" %}
+{% include "BE/cpp/ComputeJacobian" %}
+{% include "BE/cpp/UpdateTransmembranePotential" %}
+{% include "BE/cpp/ComputeOneStepExceptVoltage" %}{% include "Shared/cpp/ComputeDerivedQuantities" %}
```
Comment (jonc125) [inline chaste_codegen/templates/be_model.cpp]: Should these be on 2 lines?

File: chaste_codegen/templates/normal_model.cpp [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
+{% include "Shared/cpp/GetIntracellularCalciumConcentration" %}
 {% include "Normal/cpp/constructor" %}
 {% include "Shared/cpp/destructor" %}
-{% include "Normal/cpp/VerifyStateVariables" %}
-{% include "Normal/cpp/GetIIonic" %}
-{% include "Normal/cpp/EvaluateYDerivatives" %}{% include "Normal/cpp/ComputeDerivedQuantities" %}
-{% include "Normal/cpp/OdeSystemInformation" %}
+{% include "Shared/cpp/VerifyStateVariables" %}
+{% include "Shared/cpp/GetIIonic" %}
+{% include "Normal/cpp/EvaluateYDerivatives" %}{% include "Shared/cpp/ComputeDerivedQuantities" %}
```
Comment (jonc125) [inline chaste_codegen/templates/normal_model.cpp]: Two lines?

File: chaste_codegen/be_model.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
+import chaste_codegen as cg
+import time
+import sympy as sp
+from chaste_codegen._partial_eval import partial_eval
+from enum import Enum
+from sympy.codegen.cfunctions import (log10, log2)
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: Probably don't need brackets here :)

File: chaste_codegen/be_model.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        self._non_linear_state_vars = self._get_non_linear_state_vars()
+        self._jacobian_equations, self._jacobian_matrix = self._get_jacobian()
+        self._formatted_state_vars, self._formatted_nonlinear_state_vars, self._formatted_residual_equations, \
+            self._formatted_derivative_eqs = self._update_state_vars()
+        self._formatted_rearranged_linear_derivs, self._formatted_linear_equations = \
+            self._format_rearranged_linear_derivs()
+        self._formatted_jacobian_equations, self._formatted_jacobian_matrix_entries = self._format_jacobian()
+
+    class KINDS(Enum):
+        none = 1
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: You might want to call these `NONE`, `LINEAR`, `NONLINEAR` so the capitalisation is consistent.

File: chaste_codegen/be_model.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+            return {'state_var_index': self._state_vars.index(var),
+                    'var': self._printer.doprint(var),
+                    'g': self._printer.doprint(gh[0] if gh is not None and gh[0] is not None else 0.0),
+                    'h': self._printer.doprint(gh[1] if gh is not None and gh[1] is not None else 0.0)}
+
+        # get linear state vars
+        linear_sv = [d for d in self._y_derivatives if d.args[0] not in self._non_linear_state_vars
+                     and not d.args[0] == self._membrane_voltage_var]
+
+        # fill in the equations containing state vars only
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: I think it'd help future readers to expand on your comments explaining what this is doing!

File: chaste_codegen/be_model.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+        for eq in self._jacobian_matrix:
+            jacobian_symbols.update(eq.free_symbols)
+
+        non_linear_derivs = [eq.lhs for eq in self._derivative_equations if isinstance(eq.lhs, sp.Derivative)
+                             and eq.lhs.args[0] in self._non_linear_state_vars]
+        non_linear_eqs = self._model.get_equations_for(non_linear_derivs)
+        for eq in non_linear_eqs:
+            residual_eq_symbols.update(eq.rhs.free_symbols)
+        non_linear_eqs = [self._printer.doprint(eq.lhs) for eq in non_linear_eqs]
+        for i, d in enumerate(formatted_derivative_eqs):
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: You can do this loop without enumerate, and the ones below too I think.

File: chaste_codegen/__init__.py [hunk 11]
```diff
@@ -3,7 +3,7 @@
 """
 import logging
 from cellmlmanip.transpiler import Transpiler
-from ._math_functions import _exp, _abs, _acos, _cos, _sqrt, _sin
+from ._math_functions import exp_, abs_, acos_, cos_, sqrt_, sin_
```
[the remark this answers is not in the document]
Comment (MauriceHendrix) [reply chaste_codegen/__init__.py]: thanks for the tip of using isort

File: chaste_codegen/chaste_model.py [hunk 12]
```diff
@@ -600,11 +604,25 @@ def _get_derivative_eqs_exlc_voltage(self):
...
+        # stat with derivatives for V only and add all equations used
+        eqs = []
+        derivatives = set([deriv for deriv in self._y_derivatives if deriv.args[0] == self._membrane_voltage_var])
+        num_derivatives = -1
+        while num_derivatives < len(derivatives):
+            num_derivatives = len(derivatives)
+            eqs = [eq for eq in self._derivative_equations if eq.lhs in derivatives]
+            for eq in eqs:
+                for s in self._model.find_variables_and_derivatives([eq.rhs]):
+                    derivatives.add(s)
```
[the remark this answers is not in the document]
Comment (MauriceHendrix) [reply chaste_codegen/chaste_model.py]: fair point

File: chaste_codegen/be_model.py [hunk 13]
```diff
@@ -23,75 +26,75 @@ def __init__(self, model, file_name, **kwargs):
...
-                    result = BeModel.KINDS.Nonlinear
+                    result = BeModel.KINDS.NONLINEAR
                     break
                 elif lin == 1:
-                    result = BeModel.KINDS.Linear
+                    result = BeModel.KINDS.LINEAR
         elif isinstance(expr, sp.Add) or isinstance(expr, sp.boolalg.BooleanFunction) or\
                 isinstance(expr, sp.relational.Relational):
-            # Linear if any operand linear, and none non-linear
+            # LINEAR if any operand linear, and NONE NONLINEAR
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: Probably didn't want uppercase here :)

File: chaste_codegen/be_model.py [hunk 14]
```diff
@@ -203,7 +219,7 @@ def _get_jacobian(self):
         derivative_eq_matrix = sp.Matrix(derivative_eqs)
         jacobian_matrix = derivative_eq_matrix.jacobian(state_var_matrix)
         # update state variables
-        jacobian_equations, jacobian_matrix = sp.cse(jacobian_matrix, order='none')
+        jacobian_equations, jacobian_matrix = sp.cse(jacobian_matrix, order='NONE')
```
Comment (jonc125) [inline chaste_codegen/be_model.py]: Don't think this should change?

Comment (MauriceHendrix) [reply chaste_codegen/be_model.py]: uups lol

**Pull request merged**